### PR TITLE
refactor!: refactor shard version logic

### PR DESCRIPTION
**Pull request by ZuLiangWang**
## Rationale
Refer to this issue: https://github.com/CeresDB/ceresmeta/issues/263

## Detailed Changes
* Reconstruct the process of create/drop table so that the update of shard version depends on CeresDB

## Test Plan
Pass existing unit tests and integration tests.

### CODE REVIEW

File: server/coordinator/eventdispatch/dispatch.go [hunk 1]
```diff
@@ -11,8 +11,8 @@ import (
 type Dispatch interface {
 	OpenShard(context context.Context, address string, request OpenShardRequest) error
 	CloseShard(context context.Context, address string, request CloseShardRequest) error
-	CreateTableOnShard(context context.Context, address string, request CreateTableOnShardRequest) error
-	DropTableOnShard(context context.Context, address string, request DropTableOnShardRequest) error
+	CreateTableOnShard(context context.Context, address string, request CreateTableOnShardRequest) (error, uint64)
```
Comment (ShiKaiWi) [inline server/coordinator/eventdispatch/dispatch.go:14]: The idiom way is to put the error last in the return parameters.

File: server/cluster/metadata/types.go [hunk 2]
```diff
@@ -107,15 +107,16 @@ type CloseTableRequest struct {
...
 
 type MigrateTableRequest struct {
-	SchemaName string
-	TableNames []string
-	OldShardID storage.ShardID
-	NewShardID storage.ShardID
+	SchemaName            string
+	TableNames            []string
+	OldShardID            storage.ShardID
+	latestOldShardVersion uint64
```
Comment (ShiKaiWi) [inline server/cluster/metadata/types.go:113]: Shall we introduce a struct call VersionShard to simplify these four fields into two?

**Review by ShiKaiWi — APPROVED** [review]

File: server/cluster/metadata/types.go (hunk 2 above)
Comment (ZuLiangWang) [reply server/cluster/metadata/types.go:113]: In fact, `MigrateTableRequest` is useless now,  the migration table is not currently in actual use. I will add a todo here and modify it together with the subsequent reconstruction of the migration table.

File: server/cluster/manager.go [hunk 3]
```diff
@@ -290,13 +290,37 @@ func (m *managerImpl) GetTablesByShardIDs(clusterName, _ string, shardIDs []stor
...
-	_, err = cluster.metadata.DropTable(ctx, schemaName, tableName)
+	table, ok, err := cluster.GetMetadata().GetTable(schemaName, tableName)
+	if !ok {
+		return metadata.ErrTableNotFound
+	}
+	if err != nil {
+		return errors.WithMessage(err, "get table")
+	}
+
+	getShardNodeResult, err := cluster.GetMetadata().GetShardNodeByTableIDs([]storage.TableID{table.ID})
```
Comment (ShiKaiWi) [inline server/cluster/manager.go:309]: The metadata can be declared as one variable to avoid verbose call.

File: server/cluster/manager.go [hunk 4]
```diff
@@ -290,13 +290,37 @@ func (m *managerImpl) GetTablesByShardIDs(clusterName, _ string, shardIDs []stor
...
+	if err != nil {
+		return errors.WithMessage(err, "get shard node by tableID")
+	}
+	if len(getShardNodeResult.ShardNodes[table.ID]) != 1 || len(getShardNodeResult.Version) != 1 {
+		return metadata.ErrShardNotFound
+	}
+
+	shardID := getShardNodeResult.ShardNodes[table.ID][0].ID
+	version := getShardNodeResult.Version[shardID]
```
Comment (ShiKaiWi) [inline server/cluster/manager.go:318]: Shall we check to ensure the version won't be nil?

File: server/cluster/manager.go [hunk 5]
```diff
@@ -290,13 +290,37 @@ func (m *managerImpl) GetTablesByShardIDs(clusterName, _ string, shardIDs []stor
...
+
+	_, err = cluster.metadata.DropTable(ctx, schemaName, tableName, metadata.ShardVersionUpdate{
```
Comment (ShiKaiWi) [inline server/cluster/manager.go:320]: Is there any different between `cluster.GetMetadata()` and `cluster.metadata`?

File: server/cluster/metadata/cluster_metadata.go [hunk 6]
```diff
@@ -358,7 +358,7 @@ func (c *ClusterMetadata) DropTableMetadata(ctx context.Context, schemaName, tab
 	return dropRes, nil
 }
 
-func (c *ClusterMetadata) CreateTable(ctx context.Context, request CreateTableRequest) (CreateTableResult, error) {
+func (c *ClusterMetadata) CreateTable(ctx context.Context, request CreateTableRequest, latestShardVersion uint64) (CreateTableResult, error) {
```
Comment (ShiKaiWi) [inline server/cluster/metadata/cluster_metadata.go:361]: Shall we put the latestShardVersion inside CreateTableRequest?

File: server/cluster/metadata/cluster_metadata.go [hunk 7]
```diff
@@ -191,7 +191,7 @@ func (c *ClusterMetadata) GetShardTables(shardIDs []storage.ShardID) map[storage
 
 // DropTable will drop table metadata and all mapping of this table.
 // If the table to be dropped has been opened multiple times, all its mapping will be dropped.
-func (c *ClusterMetadata) DropTable(ctx context.Context, schemaName, tableName string) (DropTableResult, error) {
+func (c *ClusterMetadata) DropTable(ctx context.Context, schemaName, tableName string, shardVersionUpdate ShardVersionUpdate) (DropTableResult, error) {
```
Comment (ShiKaiWi) [inline server/cluster/metadata/cluster_metadata.go:194]: Shall we declare a struct called DropTableRequest carrying all the params to keep the same style with CreateTable's signature?

Comment (ShiKaiWi) [reply server/cluster/metadata/cluster_metadata.go:194]: Why ShardVersionUpdate is used here while just latest shard version is used by `CreateTable`?

File: server/coordinator/procedure/ddl/common_util.go [hunk 8]
```diff
@@ -150,24 +147,24 @@ func DispatchDropTable(ctx context.Context, clusterMetadata *metadata.ClusterMet
 		CreatedAt:     0,
 	}
 
+	var curVersion uint64
 	for _, shardNode := range shardNodes {
-		err = dispatch.DropTableOnShard(ctx, shardNode.NodeName, eventdispatch.DropTableOnShardRequest{
+		curVersion, err = dispatch.DropTableOnShard(ctx, shardNode.NodeName, eventdispatch.DropTableOnShardRequest{
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/ddl/common_util.go:152]:
```suggestion
		latestShardVersion, err = dispatch.DropTableOnShard(ctx, shardNode.NodeName, eventdispatch.DropTableOnShardRequest{
```

File: server/coordinator/procedure/ddl/droppartitiontable/create_drop_partition_table_test.go [hunk 9]
```diff
@@ -54,6 +54,7 @@ func TestCreateAndDropPartitionTable(t *testing.T) {
 		tableName := fmt.Sprintf("%s_%d", test.TestTableName0, i)
 		subTableNames := genSubTables(tableName, testSubTableNum)
 		testCreatePartitionTable(ctx, t, dispatch, c, s, shardPicker, shardNode.NodeName, tableName, subTableNames)
+		println("create partition table finish %s", tableName)
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/ddl/droppartitiontable/create_drop_partition_table_test.go:57]: Remove the debug codes.

File: server/coordinator/procedure/ddl/droppartitiontable/drop_partition_table.go [hunk 10]
```diff
@@ -397,15 +397,18 @@ func dispatchDropDataTable(req *callbackRequest, dispatch eventdispatch.Dispatch
...
-		if err := ddl.DispatchDropTable(req.ctx, clusterMetadata, dispatch, schema, table, shardVersionUpdate); err != nil {
+		latestShardVersion, err := ddl.DropTableOnShard(req.ctx, clusterMetadata, dispatch, schema, table, shardVersionUpdate)
+		if err != nil {
 			return errors.WithMessagef(err, "drop table, table:%s", tableName)
 		}
 
-		_, err = clusterMetadata.DropTable(req.ctx, req.schemaName(), tableName)
+		_, err = clusterMetadata.DropTable(req.ctx, req.schemaName(), tableName, metadata.ShardVersionUpdate{
+			ShardID:     shardID,
+			PrevVersion: latestShardVersion,
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/ddl/droppartitiontable/drop_partition_table.go:410]: It's weird to assign the lastestShardVersion to PrevVersion field? Is it reasonable?

File: server/coordinator/procedure/test/common.go [hunk 11]
```diff
@@ -53,27 +53,27 @@ const (
 
 type MockDispatch struct{}
 
-func (m MockDispatch) OpenShard(_ context.Context, _ string, _ eventdispatch.OpenShardRequest) error {
+func (m MockDispatch) OpenShard(context context.Context, address string, request eventdispatch.OpenShardRequest) error {
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/test/common.go:56]: Won't the linter complain about these unused parameters?

File: server/cluster/manager.go (hunk 5 above)
Comment (ZuLiangWang) [reply server/cluster/manager.go:320]: `cluster.GetMetadata()` is used to get `metadata` in other package, replace all `cluster.GetMetadata()` in this package to `cluster.metadata`

File: server/coordinator/procedure/ddl/common_util.go [hunk 12]
```diff
@@ -46,14 +46,14 @@ func CreateTableOnShard(ctx context.Context, c *metadata.ClusterMetadata, dispat
 		}
 	}
 	if !found {
-		return errors.WithMessagef(procedure.ErrShardLeaderNotFound, "shard node can't find leader, shardID:%d", shardID)
+		return 0, errors.WithMessagef(procedure.ErrShardLeaderNotFound, "shard node can't find leader, shardID:%d", shardID)
 	}
 
-	err = dispatch.CreateTableOnShard(ctx, leader.NodeName, request)
+	curVersion, err := dispatch.CreateTableOnShard(ctx, leader.NodeName, request)
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/ddl/common_util.go:52]: curVersion -> latestShardVersion?

File: server/cluster/metadata/topology_manager.go [hunk 13]
```diff
@@ -312,74 +296,18 @@ func (m *TopologyManagerImpl) RemoveTable(ctx context.Context, shardID storage.S
...
 	return ShardVersionUpdate{
-		ShardID:     shardID,
-		CurrVersion: prevVersion + 1,
-		PrevVersion: prevVersion,
+		ShardID:       shardID,
+		LatestVersion: latestVersion,
 	}, nil
 }
 
-func (m *TopologyManagerImpl) EvictTable(ctx context.Context, tableID storage.TableID) ([]ShardVersionUpdate, error) {
```
[the remark this answers is not in the document]
Comment (ZuLiangWang) [reply server/cluster/metadata/topology_manager.go:324]: `EvictTable` is used to delete tables distributed in multiple shards, that is, the main table of the old impementation of the partition table. Due to subsequent implementation modifications of the partition table, the main table is no longer controlled by `CeresMeta` to open, so it is deleted directly without any alternative method.

File: server/cluster/metadata/topology_manager.go [hunk 14]
```diff
@@ -274,7 +258,7 @@ func (m *TopologyManagerImpl) RemoveTable(ctx context.Context, shardID storage.S
 	if !ok {
 		return emptyUpdate, ErrShardNotFound.WithCausef("shard id:%d", shardID)
 	}
-	prevVersion := shardView.Version
+	latestVersion := shardView.Version
```
[the remark this answers is not in the document]
Comment (ZuLiangWang) [reply server/cluster/metadata/topology_manager.go:261]: I found that the data in the return value does not seem to have any effect at present, and I plan to refactor this code again.

File: server/cluster/metadata/cluster_metadata_test.go [hunk 15]
```diff
@@ -177,9 +178,14 @@ func testTableOperation(ctx context.Context, re *require.Assertions, m *metadata
 	re.Equal(storage.ShardID(1), routeResult.RouteEntries[testTableName].NodeShards[0].ShardInfo.ID)
 
 	// Drop table already created.
-	dropResult, err := m.DropTable(ctx, testSchema, testTableName)
+	dropResult, err := m.DropTable(ctx, metadata.DropTableRequest{
+		SchemaName:    testSchema,
+		TableName:     testTableName,
+		ShardID:       storage.ShardID(1),
+		LatestVersion: 0,
```
[the remark this answers is not in the document]
Comment (ZuLiangWang) [reply server/cluster/metadata/cluster_metadata_test.go:185]: I think there is no need to check the version here. The wrong version will be found in the subsequent table creation and deletion.

**Review by ShiKaiWi — APPROVED** [review]

**Pull request merged**